igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerRegistryCacheV2KeysMigration.java

```diff
@@ -131,48 +125,75 @@ private void migrateBatch(List<DockerRegistryV1Key> oldKeys) {
      * different methods that are the same thing until this is fixed.
      * TODO rz - switch to common `scan` command
      */
-    private List<DockerRegistryV1Key> getV1Keys(MultiKeyCommands client) {
-        if (redis instanceof DynomiteClientDelegate) {
-            return v1Keys((DynoJedisClient) client);
+    private int getV1Keys(MultiKeyCommands client) {
+        try {
+            if (redis instanceof DynomiteClientDelegate) {
+                return v1Keys((DynoJedisClient) client);
+            }
+            return v1Keys((Jedis) client);
+        } catch (InterruptedException e) {
+            log.error("Migration could not complete because it was interrupted", e);
+            return -1;
         }
-        return v1Keys((Jedis) client);
     }
 
-    private Function<List<String>, List<DockerRegistryV1Key>> oldKeysCallback =
-        (keys) -> keys.stream().map(this::readV1Key).filter(Objects::nonNull).collect(Collectors.toList());
+    private Function<List<String>, Integer> oldKeysCallback =
+        (keys) -> {
+            List<DockerRegistryV1Key> v1Keys = keys.stream()
+                .map(this::readV1Key)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+            int size = keys.size();
+
+            log.info("Migrating {} v1 keys", size);
+            migrateBatch(v1Keys);
+
+            return size;
+        };
 
     /**
      * Dynomite-compat v1keys
      */
-    private List<DockerRegistryV1Key> v1Keys(DynoJedisClient dyno) {
-        List<DockerRegistryV1Key> keys = new ArrayList<>();
+    private int v1Keys(DynoJedisClient dyno) throws InterruptedException {
+        int numMigrated = 0;
 
         String pattern = oldIndexPattern();
-        CursorBasedResult<String> result;
+        CursorBasedResult<String> result = null;
         do {
-            result = dyno.dyno_scan(pattern);
-            keys.addAll(oldKeysCallback.apply(result.getResult()));
-        } while (!result.isComplete());
+            try {
+                result = dyno.dyno_scan(pattern);
+                numMigrated += oldKeysCallback.apply(result.getResult());
+            } catch (DynoException e) {
```

Comment:
Should we keep retrying forever? Are there some exceptions that mean we don't want to retry?

Reply:
Good point. I've added a failure counter to abort if there's an excessive amount of consecutive failures.